### PR TITLE
Implement the dual mode for Thanos Receiver

### PR DESCRIPTION
**Pull request by yashrsharma44**
Implementation PR for the Thanos Split Receiver proposal - https://thanos.io/tip/proposals/202012_receive_split.md/

**PR is incomplete, pushed for tracking changes and quick review**
TODO: 
- [ ] Define behaviour when one mode is enabled, and other mode's behaviour is accessed
- [ ] Tests and Docs

Signed-off-by: Yash Sharma <yashrsharma44@gmail.com>

### CODE REVIEW

File: cmd/thanos/receive.go [hunk 1]
```diff
@@ -131,16 +129,22 @@ func registerReceive(app *extkingpin.App) {
 			AllowOverlappingBlocks: *tsdbAllowOverlappingBlocks,
 		}
 
-		// Local is empty, so try to generate a local endpoint
-		// based on the hostname and the listening port.
+		// Make sure the both localEndpoint and (hashringsFilePath/hashringsFileContent) are not empty
+		if *localEndpoint == "" && *hashringsFilePath == "" && *hashringsFileContent == "" {
+			return errors.New("both the receive.local-endpoint and (receive.hashrings-file/receive.hashrings) are empty, either of them should be non-empty.")
```
Comment (bwplotka) [inline cmd/thanos/receive.go:134]: Can we refer to docs?

File: cmd/thanos/receive.go [hunk 2]
```diff
@@ -226,9 +231,21 @@ func runReceive(
...
 	comp component.SourceStoreAPI,
 	hashFunc metadata.HashFunc,
+	receiverMode receive.ReceiverMode,
 ) error {
+
 	logger = log.With(logger, "component", "receive")
 	level.Warn(logger).Log("msg", "setting up receive")
+
+	switch receiverMode {
+	case receive.INGESTOR_MODE:
```
Comment (bwplotka) [inline cmd/thanos/receive.go:241]: I wonder if we can use this time group together distributing work and ingesting mode, so it's selected in one time. (:

Comment (bwplotka) [reply cmd/thanos/receive.go:241]: let's start with separate functions to register all?

File: cmd/thanos/receive.go (hunk 1 above)
Comment (yashrsharma44) [reply cmd/thanos/receive.go]: Done

File: cmd/thanos/receive.go (hunk 2 above)
Comment (yashrsharma44) [reply cmd/thanos/receive.go:241]: I tried using functions to separate the logic but seems like the functions convoluted the codebase, and passing a large list of parameters was distracting to read.

I have tried collecting the similar sections into one `if` condition, let me know if this makes sense (:

File: cmd/thanos/receive.go [hunk 3]
```diff
@@ -131,16 +129,22 @@ func registerReceive(app *extkingpin.App) {
 			AllowOverlappingBlocks: *tsdbAllowOverlappingBlocks,
 		}
 
-		// Local is empty, so try to generate a local endpoint
-		// based on the hostname and the listening port.
+		// Make sure the both localEndpoint and (hashringsFilePath/hashringsFileContent) are not empty
```
Comment (squat) [inline cmd/thanos/receive.go:132]: This seems wrong to me as it makes it impossible for a node to be in "ingestor" mode. It's absolutely fine for a Receive node to have no local endpoint and no hashring configured. This is the default way the receiver runs: all data it receives is deemed to be for itself and is stored in a local TSDB.

File: cmd/thanos/receive.go [hunk 4]
```diff
@@ -131,16 +129,22 @@ func registerReceive(app *extkingpin.App) {
...
+		if *localEndpoint == "" && *hashringsFilePath == "" && *hashringsFileContent == "" {
+			return errors.New("both the receive.local-endpoint and (receive.hashrings-file/receive.hashrings) are empty; see https://thanos.io/tip/components/receive.md for details.")
+		}
+
+		// Default behavior for receiver is both ingestion and distribution
+		receiverMode := receive.DUAL_MODE
```
Comment (squat) [inline cmd/thanos/receive.go:138]: I find this abstraction a bit more confusing than it is helpful as it forces us to have lots of checks in the code like "if this but not this" or "if not that but yes this". I think it would be more helpful to have one single flag that says "i run a TSDB". That is really the only thing we care about. Even if there is no hashring file configured, the ingestor mode MUST run a gRPC server because this mode might be at the bottom of a routing tree and know nothing about the distributor nodes that are forwarding traffic to it. This means that there is no difference at all between "dual mode" and "ingestor"

Comment (yashrsharma44) [reply cmd/thanos/receive.go:138]: Okay, so I think, we can introduce a flag, `disableTSDB`, which would instruct the receiver to **only route data**, and not ingest data. 
So this would be the following behavior of the receiver - 

* **Both Ingestion and Distribution** - Default behavior of receiver, works as it used to, no extra flags.
* **Only Ingestion** - Just don't pass hash ring file path/content, the receiver would just ingest data.
* **Only Distribution** - We can create a flag, `disableTSDB`, which would stop creating a local tsdb for ingesting data.

Thoughts?

File: cmd/thanos/receive.go (hunk 3 above)
Comment (yashrsharma44) [reply cmd/thanos/receive.go:132]: I agree with your thoughts, we currently try to fetch the host as our local endpoint, so this behavior has removed that functionality. 

Originally, I thought that we should explicitly pass the `localEndpoint`, but now with the [other approach](https://github.com/thanos-io/thanos/pull/4231#discussion_r632469132), we don't have to change the behavior as such, and I can totally remove this check!

File: cmd/thanos/receive.go (hunk 4 above)
Comment (squat) [reply cmd/thanos/receive.go]: I don't see a need for this flag. We can unambiguously infer everything we need to know with the presence/absence of the current flags:

1. Hashring info != "" && local endpoint == "": means only in distribution mode, i.e. don't start a TSDB
2. Everything else means start a regular receiver with TSDB, http server, and a gRPC server

This covers all of the possible cases we could want, does not require any new flags, and keeps things simpler IMO

Comment (yashrsharma44) [reply cmd/thanos/receive.go]: Sure, we can configure this decision internally also, Thanks!

**Comment by yashrsharma44** [comment]
Hi @squat !

I have added an e2e test scenario about creating 3 ingestor(hashrings contain only the local endpoint) and 1 distributor. Can you check if the logic for the same is implemented correctly?

I ran the tests, and cross-checked if the behaviour is correct, but would be nice to have your opinion as well 💯

File: cmd/thanos/receive.go [hunk 5]
```diff
@@ -132,15 +130,18 @@ func registerReceive(app *extkingpin.App) {
...
-		if *localEndpoint == "" {
-			hostname, err := os.Hostname()
-			if hostname == "" || err != nil {
-				return errors.New("--receive.local-endpoint is empty and host could not be determined.")
-			}
-			parts := strings.Split(*grpcBindAddr, ":")
-			port := parts[len(parts)-1]
-			*localEndpoint = net.JoinHostPort(hostname, port)
-		}
+		// if *localEndpoint == "" {
```
Comment (yashrsharma44) [inline cmd/thanos/receive.go:133]: Do you want me to uncomment this part? However, we might not be able to support the distributor mode, as the `localEndpoint` would always have a non-nil value.

**Review by kakkoyun — COMMENTED** [review]
This is looking however it gives me an unsettling feeling. It's really hard to follow up on changes and modes.

Could we try to encapsulate logic into functions to increase readability?

I know this was an existing issue but this is a chance to improve the control flow.

File: cmd/thanos/receive.go (hunk 5 above)
Comment (kakkoyun) [reply cmd/thanos/receive.go:133]: We should remove this if it's not used.

**Comment by yashrsharma44** [comment]
> I know this was an existing issue but this is a chance to improve the control flow.

Can you link the issue?

File: cmd/thanos/receive.go (hunk 4 above)
Comment (yashrsharma44) [reply cmd/thanos/receive.go]: In the distribution mode, do we need to start the gRPC server? 

Also, can you list out which parts are enabled in the ingestion / distribution mode
cc @squat

**Comment by yashrsharma44** [comment]
> > I know this was an existing issue but this is a chance to improve the control flow.
> 
> Can you link the issue?

Tracking issue - #2248

File: cmd/thanos/receive.go (hunk 4 above)
Comment (yashrsharma44) [reply cmd/thanos/receive.go]: Sounds great!

File: cmd/thanos/receive.go [hunk 6]
```diff
@@ -214,235 +214,302 @@ func runReceive(
...
+	dbReady chan struct{},
+	comp component.SourceStoreAPI,
+	dbs *receive.MultiTSDB,
+	webHandler *receive.Handler,
+	grpcLogOpts []grpc_logging.Option,
+	tagOpts []tags.Option,
+	grpcProbe *prober.GRPCProbe,
+
+) error {
+	g.Add(func() error {
```
Comment (kakkoyun) [inline cmd/thanos/receive.go:289]: This looks it does the job. I'd prefer keeping the `run.Group` in the top-level function and extract whatever we pass to `g.Add` on a different function.

File: test/e2e/receive_test.go [hunk 7]
```diff
@@ -56,6 +56,77 @@ func ErrorHandler(_ http.ResponseWriter, _ *http.Request, err error) {
 
 func TestReceive(t *testing.T) {
 	t.Parallel()
+	// t.Run("receive_distributor_ingestor_mode", func(t *testing.T) {
```
Comment (kakkoyun) [inline test/e2e/receive_test.go:59]: Why these are commented? Should we mark this PR as WIP?

File: cmd/thanos/receive.go [hunk 8]
```diff
@@ -214,235 +214,302 @@ func runReceive(
...
+			startGRPC <- struct{}{}
+		}
+		if s != nil {
+			s.Shutdown(err)
+		}
+		return nil
+	}, func(error) {})
+	// We need to be able to start and stop the gRPC server
+	// whenever the DB changes, thus it needs its own run group.
+	g.Add(func() error {
```
Comment (kakkoyun) [inline cmd/thanos/receive.go:328]: The function name says `setupGRPCServer` but this also starts servers. It'd be good to scope of functions small as possible.

**Review by kakkoyun — COMMENTED** [review]
This is on a good path. Thanks a lot @yashrsharma44.

I have added comments inline.

File: test/e2e/receive_test.go (hunk 7 above)
Comment (yashrsharma44) [reply test/e2e/receive_test.go:59]: I wanted to check if the original tests pass after refactoring. I haven't added the logic for split behaviour, so once the tests pass, I will uncomment them.

File: cmd/thanos/receive.go (hunk 8 above)
Comment (yashrsharma44) [reply cmd/thanos/receive.go:328]: Shall we split up into two small functions?

File: cmd/thanos/receive.go (hunk 5 above)
Comment (yashrsharma44) [reply cmd/thanos/receive.go]: Removed. This is a breaking change, so we should mention this in the changelog

File: cmd/thanos/receive.go (hunk 8 above)
Comment (yashrsharma44) [reply cmd/thanos/receive.go]: Done

File: test/e2e/receive_test.go (hunk 7 above)
Comment (yashrsharma44) [reply test/e2e/receive_test.go]: Added the tests back

File: cmd/thanos/receive.go (hunk 5 above)
Comment (yashrsharma44) [reply cmd/thanos/receive.go]: **Yup, this is a breaking change, so lets make a note for this**

File: cmd/thanos/receive.go [hunk 9]
```diff
@@ -203,246 +204,364 @@ func runReceive(
...
+			},
+		)
+	}
+
+	level.Info(logger).Log("msg", "starting receiver")
+	return nil
+}
+
+// runGRPCServer starts the grpc server, once it receives a signal from the startGRPCListening channel.
+func runGRPCServer(g *run.Group,
```
[the remark this answers is not in the document]
Comment (yashrsharma44) [reply cmd/thanos/receive.go]: @muse-dev ignore

File: cmd/thanos/receive.go (hunk 8 above)
Comment (yashrsharma44) [reply cmd/thanos/receive.go:337]: For some reason, splitting up tasks are creating a nil pointer for the grpcServer, so I reverted back the change. Let me know if you have any suggestion

**Review by vovanloc0798 — APPROVED** [review]

**Comment by yashrsharma44** [comment]
I might set up this branch and play with the benchmarks using this PR - https://github.com/thanos-io/thanosbench/pull/34

So will report the statistics and benchmarks, once I run them up 😎

File: cmd/thanos/receive.go [hunk 10]
```diff
@@ -74,17 +72,9 @@ func registerReceive(app *extkingpin.App) {
...
-		if conf.endpoint == "" {
-			hostname, err := os.Hostname()
-			if hostname == "" || err != nil {
-				return errors.New("--receive.local-endpoint is empty and host could not be determined.")
-			}
-			parts := strings.Split(*conf.grpcBindAddr, ":")
-			port := parts[len(parts)-1]
-			conf.endpoint = net.JoinHostPort(hostname, port)
-		}
+		// enable ingestion if endpoint is specified, or both the hashrings config are empty,
```
Comment (squat) [inline cmd/thanos/receive.go:75]: Just a nit: let's break up the sentences to make sure that the logic in the comment is crystal clear:
```suggestion
		// Enable ingestion if endpoint is specified or if both the hashrings configs are empty.
		// Otherwise, run the receiver exclusively as a distributor.
```

File: cmd/thanos/receive.go [hunk 11]
```diff
@@ -203,246 +204,339 @@ func runReceive(
...
+func setupHashring(g *run.Group,
+	logger log.Logger,
+	reg *prometheus.Registry,
+	conf *receiveConfig,
+	hashringChangedChan chan struct{},
+	webHandler *receive.Handler,
+	statusProber prober.Probe,
+	reloadGRPCServer chan struct{},
+	enableIngestion bool,
+
```
Comment (squat) [inline cmd/thanos/receive.go:362]: nit
```suggestion
```

File: test/e2e/e2ethanos/services.go [hunk 12]
```diff
@@ -242,6 +243,50 @@ func NewReceiver(sharedDir string, networkName string, name string, replicationF
 	return receiver, nil
 }
 
+func NewReceiverWithDistributorMode(sharedDir string, networkName string, name string, replicationFactor int, hashring ...receive.HashringConfig) (*Service, error) {
```
Comment (squat) [inline test/e2e/e2ethanos/services.go:246]: Naming nit: there is no concrete thing as "distribution" mode, but we can talk concretely about whether the receiver supports ingesting to a TSDB or not. I would propose calling this `NewReceiverWithoutTSDB`

Comment (squat) [reply test/e2e/e2ethanos/services.go:246]: But not a hard blocker here

**Review by squat — DISMISSED** [review]
Overall looks great to me! I think the PR is a lot simpler now without the extra concepts of the different "modes" :))

File: test/e2e/e2ethanos/services.go (hunk 12 above)
Comment (yashrsharma44) [reply test/e2e/e2ethanos/services.go:246]: Lets add this to the suggestions, should be consistent with the nomenclature :)

Comment (yashrsharma44) [reply test/e2e/e2ethanos/services.go]: Done

**Review by kakkoyun — APPROVED** [review]
:1st_place_medal: :rocket:
